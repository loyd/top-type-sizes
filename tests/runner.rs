use structopt::StructOpt;

use top_type_sizes::*;

fn test(content: &'static str) {
    let refined_content = reader::read(content.as_bytes()).unwrap();

    let types = parser::parse(&refined_content).unwrap();
    insta::with_settings!({ description => "internal representation" }, {
        insta::assert_yaml_snapshot!("types", types);
    });

    snap_output(&types, &[]);
    snap_output(&types, &["-ws"]);
    snap_output(&types, &["-h8"]);
    snap_output(&types, &["-ws", "-h16"]);
}

fn snap_output(types: &[schema::Type], cmd: &[&str]) {
    let full_cmd = std::iter::once(&"top-type-sizes")
        .chain(cmd)
        .copied()
        .collect::<Vec<_>>();

    let options = options::Options::from_iter(&full_cmd);
    let snap_name = {
        let mut parts = cmd.to_vec();
        parts.insert(0, "output");
        parts.join("")
    };

    let types = transformer::transform(types.to_vec(), &options);
    let output = formatter::format(types, &options);

    insta::with_settings!({ description => full_cmd.join(" ") }, {
        insta::assert_snapshot!(snap_name, output);
    })
}

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            insta::with_settings!(
            {
                snapshot_path => concat!("snapshots/", stringify!($name)),
                prepend_module_to_snapshot => false,
                omit_expression => true,
            },
            {
                test(include_str!(concat!("samples/", stringify!($name), ".txt")))
            })
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
