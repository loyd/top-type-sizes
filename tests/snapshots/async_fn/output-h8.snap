---
source: tests/runner.rs
description: top-type-sizes -h8
---
16386 [async fn body@$DIR/async.rs:8:36: 11:2] align=1
  16385 variant Suspend0
       8192 arg (upvar) align=1 offset=0
       8192 arg
   8192 variant Unresumed, Returned, Panicked
       8192 arg (upvar) align=1 offset=0

8192 std::mem::MaybeUninit<[u8; 8192]> align=1
   8192 variant MaybeUninit
       8192 value

8192 std::mem::ManuallyDrop<[u8; 8192]> align=1
   8192 value

1 std::task::Poll<()> align=1
      0 variant Ready, Pending

1 std::mem::MaybeUninit<[async fn body@$DIR/async.rs:6:17: 6:19]> align=1
      1 variant MaybeUninit

1 std::mem::ManuallyDrop<[async fn body@$DIR/async.rs:6:17: 6:19]> align=1

1 [async fn body@$DIR/async.rs:6:17: 6:19] align=1
      0 variant Unresumed, Returned, Panicked
